Library/Homebrew/sbom.rb

```diff
@@ -0,0 +1,361 @@
+        :bottle       => formula.bottle_hash,
+        :stable       => {
+          :version  => formula.stable&.version,
+          :url      => formula.stable&.url,
+          :checksum => formula.stable&.checksum,
+        },
+      },
+    }
+
+    # We can only get `tap_git_head` if the tap is installed locally
+    attributes[:source][:tap_git_head] = T.must(formula.tap).git_head if formula.tap&.installed?
+
+    new(attributes)
+  end
+
+  sig { params(attributes: Hash).void }
+  def initialize(attributes = {})
+    attributes.each { |key, value| instance_variable_set(:"@#{key}", value) }
+  end
+
+  sig { returns(T::Boolean) }
+  def valid?
+    data = to_spdx_sbom
+
+    schema_string, _, status = Utils::Curl.curl_output(SCHEMA)
+
+    opoo "Failed to fetch schema!" unless status.success?
+
+    Homebrew.install_bundler_gems!(groups: ["bottle"])
+    require "json_schemer"
+
+    schemer = JSONSchemer.schema(schema_string)
+
+    return true if schemer.valid?(data)
+
+    opoo "SBOM is not valid, not writing to disk!"
+    schemer.validate(data).to_a.each do |error|
+      odebug error["error"]
+    end
+
+    false
+  end
+
+  sig { void }
+  def write
+    # If this is a new installation, the cache of installed formulae
+    # will no longer be valid.
+    Formula.clear_cache unless spdxfile.exist?
+
+    self.class.cache[spdxfile] = self
+
+    return unless valid?
+
+    spdxfile.atomic_write(JSON.pretty_generate(to_spdx_sbom))
+  end
+
+  sig { params(runtime_dependency_declaration: T::Array[Hash], compiler_declaration: Hash).returns(T::Array[Hash]) }
+  def generate_relations_json(runtime_dependency_declaration, compiler_declaration)
+    runtime = runtime_dependency_declaration.map do |dependency|
+      {
+        :spdxElementId      => dependency[:SPDXID],
+        :relationshipType   => :RUNTIME_DEPENDENCY_OF,
+        :relatedSpdxElement => "SPDXRef-Bottle-#{name}",
+      }
+    end
+    patches = source[:patches].map do |_patch|
+      {
+        :spdxElementId      => "SPDXRef-Patch-#{name}",
+        :relationshipType   => :PATCH_APPLIED,
+        :relatedSpdxElement => "SPDXRef-Archive-#{name}-src",
+      }
+    end
+
+    base = [
+      {
+        :spdxElementId      => "SPDXRef-File-#{name}",
+        :relationshipType   => :PACKAGE_OF,
+        :relatedSpdxElement => "SPDXRef-Archive-#{name}-src",
+      },
+      {
+        :spdxElementId      => "SPDXRef-Compiler",
+        :relationshipType   => :BUILD_TOOL_OF,
+        :relatedSpdxElement => "SPDXRef-Package-#{name}-src",
+      },
+    ]
+
+    if compiler_declaration["SPDXRef-Stdlib"].present?
+      base += {
+        :spdxElementId      => "SPDXRef-Stdlib",
+        :relationshipType   => :DEPENDENCY_OF,
+        :relatedSpdxElement => "SPDXRef-Bottle-#{name}",
+      }
+    end
+
+    runtime + patches + base
+  end
+
+  sig { params(runtime_dependency_declaration: T::Array[Hash], compiler_declaration: Hash).returns(T::Array[Hash]) }
```

Comment:
That would be like this, is that not too verbose?
```suggestion
  sig { params(runtime_dependency_declaration: T::Array[Hash], compiler_declaration: Hash).returns(T::Array[T::Hash[Symbol, T.any(String, T::Array[T::Hash[Symbol, String]])]]) }
```

Reply:
Yeh, that's better 👍🏻